privacy_evaluator/attacks/membership_inference/membership_inference_analysis.py

```diff
@@ -0,0 +1,129 @@
+from .data_structures.slicing import Slice
+from sklearn import metrics
+from textwrap import indent
+
+
+@dataclass
+class MembershipInferenceAttackAnalysisSliceResult:
+    """Result of the membership inference attack analysis for a single slice."""
+
+    # The slice for which this result was produced.
+    slice: Slice
+
+    # Advantage score calculated by the membership inference attack analysis.
+    advantage: float
+
+    def __str__(self):
+        """Human-readable representation of the result."""
+
+        return "\n".join(
+            (
+                "MembershipInferenceAttackAnalysisSliceResult(",
+                indent(str(self.slice), "  "),
+                f"  advantage: {self.advantage:.4f}",
+                ")",
+            )
+        )
+
+
+class MembershipInferenceAttackAnalysis:
+    """Represents the membership inference attack analysis class."""
+
+    def __init__(
+        self, attack_type: Type[MembershipInferenceAttack], input_data: AttackInputData
+    ) -> None:
+        """Initializes a MembershipInferenceAttackAnalysis class.
+
+        :param attack_type: Type of membership inference attack to analyse.
+        :param input_data: Data for the membership inference attack.
+        """
+        self.attack_type = attack_type
+        self.input_data = input_data
+
+    def analyse(
+        self,
+        target_model: Classifier,
+        x: np.ndarray,
+        y: np.ndarray,
+        membership: np.ndarray,
+        slicing: Slicing = Slicing(entire_dataset=True),
+    ) -> Iterable[MembershipInferenceAttackAnalysisSliceResult]:
+        """Runs the membership inference attack and calculates attacker's advantage for each slice.
+
+        :param target_model: Target model to attack.
+        :param x: Input data to attack.
+        :param y: True labels for `x`.
+        :param membership: Labels representing the membership for each data sample in `x`. 1 for member and 0 for non-member.
+        :param slicing: Slicing specification. The slices will be created according to the specification and the attack will be run on each slice.
+        """
+
+        # Instantiate an object of the given attack type.
+        attack = self.attack_type(
+            target_model=target_model,
+            x_train=self.input_data.x_train,
+            y_train=self.input_data.y_train,
+            x_test=self.input_data.x_test,
+            y_test=self.input_data.y_test,
+        )
+        attack.fit()
+
+        results = []
+        for slice in slices(x, y, target_model, slicing):
+            membership_prediction = attack.attack(x[slice.indices], y[slice.indices])
+
+            # Calculate the advantage score as in tensorflow privacy package.
+            tpr, fpr, _ = metrics.roc_curve(
+                membership[slice.indices],
+                membership_prediction,
+                drop_intermediate=False,
+            )
+            advantage = max(np.abs(tpr - fpr))
```

Comment:
I would suggest putting both metrics into our metrics module (even though you just use the `roc_curve` from `sklearn`) and you maybe don't even need the `roc_curve` if you only need the `tpr` and `fpr` values.

Reply:
Makes sense, I will do that.